core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestSupport.java

```diff
@@ -192,20 +187,62 @@ public void init() {
                             .setTest(true)
                             .setAuxiliaryApplication(true)
                             .setHostApplicationIsTestOnly(devModeType == DevModeType.TEST_ONLY)
-                            .setProjectRoot(Paths.get(module.getProjectDirectory()))
-                            .setApplicationRoot(PathList.from(paths))
+                            .setProjectRoot(projectDir)
+                            .setApplicationRoot(getRootPaths(module, mainModule))
                             .clearLocalArtifacts();
+
+                    final QuarkusClassLoader ctParentFirstCl;
+                    final Mode currentMode = curatedApplication.getQuarkusBootstrap().getMode();
+                    // in case of quarkus:test the application model will already include test dependencies
+                    if (Mode.CONTINUOUS_TEST != currentMode && Mode.TEST != currentMode) {
+                        // In this case the current application model does not include test dependencies.
+                        // 1) we resolve an application model for test mode;
+                        // 2) we create a new CT base classloader that includes parent-first test scoped dependencies
+                        // so that they are not loaded by augment and base runtime classloaders.
+                        var appModelFactory = curatedApplication.getQuarkusBootstrap().newAppModelFactory();
+                        appModelFactory.setBootstrapAppModelResolver(null);
+                        appModelFactory.setTest(true);
+                        appModelFactory.setLocalArtifacts(Set.of());
+                        if (!mainModule) {
+                            appModelFactory.setAppArtifact(null);
+                            appModelFactory.setProjectRoot(projectDir);
+                        }
+                        final ApplicationModel testModel = appModelFactory.resolveAppModel().getApplicationModel();
+                        bootstrapConfig.setExistingModel(testModel);
+
+                        QuarkusClassLoader.Builder clBuilder = null;
```

Comment:
This is not really proper 'parent first', they are still sitting in a ClassLoader that is not the JDK class loader.

Generally parent first has been used to hack around bugs in code that expects it to be in the same CL as the JDK itself, so I am not sure how much this actually helps?

It seems like a more correct solution would be to always put parent first test scoped deps onto the class path, but then 'hide' them from the base runtime and augment CLs. Technically other parent first classes could still see them, but I think that is unlikely to cause problems.

Reply:
It does help the pact extension case. As to the always adding test scoped parent-first dependencies, that means always resolving test scope dependencies, which we don't want to do, since it will influence version convergence with the prod dependency tree.
To do it properly, we need to completely isolate the classpath of test apps when launching in dev mode.

Reply:
Do you have an example of test scoped deps affecting non-test convergence? This seems like a problem as it would mean that tests may not be running against the same version as the actual application?

Either way I think this is fine for now, in general we should be trying to have as few parent first things as possible.

Reply:
There are instructions to reproduce the issue in Holly's issue report, in case you are curious. But, I agree, this is not a complete fix.

Reply:
I don't have a specific example but chances are this is happening more often than people realize. The same can be said about build time dependencies, such as `provided` and optional.